### PR TITLE
chore(blackboxing): designer

### PR DESCRIPTION
**Pull request by acateland**
**Please check if the PR fulfills these requirements**

- [ ] The commit(s) message(s) follows our
  [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**
Created an api around the port datastructure so client don't have to manipulate it or use it directly

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration
path for existing applications: ...

**Other information**:

### CODE REVIEW

File: src/api/port.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+import Immutable from 'immutable';
+import { PortRecord, PositionRecord } from '../constants/flowdesigner.model';
+import { PORT_SOURCE, PORT_SINK } from '../constants/flowdesigner.constants';
+
+const positionSelector = ['graphicalAttributes', 'position'];
+const componentTypeSelector = ['graphicalAttributes', 'portType'];
+const portTopologySelector = ['graphicalAttributes', 'properties', 'type'];
+const indexSelector = ['graphicalAttributes', 'properties', 'index'];
+
+export function isPositionRecord(position, doThrow) {
```
Comment (sgendre) [inline src/api/port.js]: can you provide a default value to "doThrow" like you did in the next function "isPortRecord"?

File: src/api/port.js [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+ * @param {*} port
+ * @returns {bool}
+ * @throws
+ */
+export function isPortRecordElseThrow(port) {
+	return isPortRecord(port, true);
+}
+
+/**
+ * Check if the typologie is one of the two accepted value
```
Comment (sgendre) [inline src/api/port.js]: typo on typologie ? typology or topology ?

File: src/api/port.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+ * @param {*} typology
+ * @param {bool} doThrow
+ */
+export function isTypology(typology, doThrow) {
```
Comment (sgendre) [inline src/api/port.js]: can you provide a default value to "doThrow" like you did in the previous function "isPortRecord"?

File: src/api/port.js [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	return false;
+}
+
+/**
+ * @param {string}
+ * @param {PortRecord}
+ * @returns {PortRecord}
+ */
+const setId = curry((id, port) => {
+	if (isPortRecord(port) && typeof id === 'string') {
```
Comment (sgendre) [inline src/api/port.js]: simplest condition might be located first

File: src/api/port.js [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	return false;
+}
+
+/**
+ * @param {string} nodeId
+ * @param {PortRecord} port
+ * @returns {PortRecord}
+ */
+export const setNodeId = curry((nodeId, port) => {
+	if (isPortRecord(port, true) && typeof nodeId === 'string') {
```
Comment (sgendre) [inline src/api/port.js]: simplest condition might be located first

File: src/api/port.js [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	return false;
+}
+
+/**
+ * @param {number} index
+ * @param {PortRecord} port
+ * @returns {PortRecord}
+ */
+export const setIndex = curry((index, port) => {
+	if (isPortRecord(port, true) && typeof index === 'number') {
```
Comment (sgendre) [inline src/api/port.js]: simplest condition might be located first

**Comment by sgendre** [comment]
can you expose your api object within the index.js ?

File: src/api/port.js (hunk 4 above)
Comment (acateland) [reply src/api/port.js]: is it regarding perf or code style wise ?

File: yarn-error.log [hunk 7]
```diff
@@ -0,0 +1,5687 @@
+Arguments:
```
Comment (talend-glorieux) [inline yarn-error.log]: yarn-error.log shouldn't be commited.

File: src/api/port.js [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+	}
+	return false;
+}
+
+/**
+ * @param {PortRecord} port
+ * @returns {string}
+ */
+export function getId(port) {
+	const whatver = isPortRecordElseThrow(port);
```
Comment (talend-glorieux) [inline src/api/port.js]: `whatver`? What is this?

Comment (talend-glorieux) [reply src/api/port.js]: Can't you simply remove it by putting it inside the `if` statement.

File: src/api/port.js [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+	return false;
+}
+
+/**
+ * @param {string} typology
+ * @param {PortRecord} port
+ * @returns {PortRecord}
+ */
+export const setTypology = curry((typology, port) => {
+	if (isPortRecord(port, true && isTypology(typology))) {
```
Comment (talend-glorieux) [inline src/api/port.js]: I guess you can remove that `true` statement.

File: src/api/port.js (hunk 8 above)
Comment (acateland) [reply src/api/port.js]: was a forgotten test thanks !

File: src/api/port.js (hunk 9 above)
Comment (acateland) [reply src/api/port.js]: wrongly placed parenthèsis

File: src/api/port.js (hunk 4 above)
Comment (sgendre) [reply src/api/port.js]: perf

**Comment by acateland** [comment]
@talend-glorieux yes indeed, these function check are exposed to the end user and he may not expect them to throw on simple datastructure check.

**Pull request merged**